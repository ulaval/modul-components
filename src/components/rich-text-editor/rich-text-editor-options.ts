import { library } from '@fortawesome/fontawesome';
import faAngleLeft from '@fortawesome/fontawesome-free-solid/faAngleLeft';
import faArrowLeft from '@fortawesome/fontawesome-free-solid/faArrowLeft';
import faBold from '@fortawesome/fontawesome-free-solid/faBold';
import faCompress from '@fortawesome/fontawesome-free-solid/faCompress';
import faEdit from '@fortawesome/fontawesome-free-solid/faEdit';
import faExpand from '@fortawesome/fontawesome-free-solid/faExpand';
import faExternalLinkAlt from '@fortawesome/fontawesome-free-solid/faExternalLinkAlt';
import faIndent from '@fortawesome/fontawesome-free-solid/faIndent';
import faItalic from '@fortawesome/fontawesome-free-solid/faItalic';
import faLink from '@fortawesome/fontawesome-free-solid/faLink';
import faListOl from '@fortawesome/fontawesome-free-solid/faListOl';
import faListUl from '@fortawesome/fontawesome-free-solid/faListUl';
import faOutdent from '@fortawesome/fontawesome-free-solid/faOutdent';
import faParagraph from '@fortawesome/fontawesome-free-solid/faParagraph';
import faPlus from '@fortawesome/fontawesome-free-solid/faPlus';
import faSubscript from '@fortawesome/fontawesome-free-solid/faSubscript';
import faSuperscript from '@fortawesome/fontawesome-free-solid/faSuperscript';
import faTimes from '@fortawesome/fontawesome-free-solid/faTimes';
import faTrash from '@fortawesome/fontawesome-free-solid/faTrash';
import faUnlink from '@fortawesome/fontawesome-free-solid/faUnlink';

library.add(faExpand, faCompress, faBold, faItalic, faSubscript, faSuperscript, faParagraph, faListOl, faListUl, faOutdent, faIndent, faLink, faExternalLinkAlt, faEdit, faUnlink, faTrash,
    faTimes, faArrowLeft, faPlus, faAngleLeft);

export abstract class MRichTextEditorDefaultOptions {
    public immediateVueModelUpdate: boolean = true;
    public iconsTemplate: string = 'font_awesome_5';
    public charCounterCount: boolean = false;
    public tableInsertHelper: boolean = false;
    public lineBreakerTags: string [] = [];
    public zIndex: number = 200;
    public toolbarSticky: boolean = true;
    public scrollableContainer: string = 'body';
    public toolbarStickyOffset: number = 0;
    public pluginsEnabled: string[] = ['align', 'draggable', 'embedly', 'entities', 'file', 'fontFamily', 'fontSize', 'fullscreen', 'inlineStyle', 'lineBreaker', 'link', 'lists', 'paragraphFormat', 'paragraphStyle', 'quote', 'save', 'specialCharacters', 'table', 'url', 'wordPaste', 'stylesSubMenu', 'listesSubMenu', 'insertionsSubMenu'];
    public wordPasteModal: boolean = true;

    constructor(public key: string, public language: string = 'en_CA') {
        if (!key) {
            throw new Error('In order to use the rich-text-editor you need to provide a valid froala key.');
        }
    }
}

export class MRichTextEditorStandardOptions extends MRichTextEditorDefaultOptions {
    public toolbarButtons: string[] = ['hide', 'styles-sub-menu', 'bold', 'italic', 'subscript', 'superscript', '|', 'listes-sub-menu', 'formatUL',
        'formatOL', 'outdent', 'indent', '|', 'insertLink', 'specialCharacters', '|', 'fullscreen'];

    public linkEditButtons: string[] = ['linkOpen', 'linkEdit', 'linkRemove'];

    public linkInsertButtons: string[] = [];

    constructor(key: string, language?: string | undefined) { super(key, language); }
}
