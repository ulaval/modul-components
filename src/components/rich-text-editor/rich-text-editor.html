<div class="m-rich-text"
     :class="{ 'm--is-disabled' : isDisabled,
    'm--is-waiting' : isWaiting,
    'm--has-error' : hasError,
    'm--is-valid' : isValid,
    'm--is-focus' : isFocus,
    'm--has-label': hasLabel,
    'm--has-validation-message': hasValidationMessage  }"
     @click="onClick"
     :style="{ width: inputWidth, maxWidth: inputMaxWidth }">
    <m-input-style :disabled="isDisabled"
                   :waiting="isWaiting"
                   :error="hasError"
                   :valid="isValid"
                   :label="label"
                   :label-for="id"
                   :focus="isFocus"
                   :empty="isEmpty"
                   :border-top="!readonly"
                   :required-marker="requiredMarker"
                   :tag-style="tagStyle"
                   :readonly="readonly">
        <froala :config="internalOptions"
                :value="value"
                @input="refreshModel"
                @focus="onFocus"
                @blur="onBlur"
                @change="onChange"
                @keypress.enter.prevent
                @keyup="onKeyup"
                @keydown="onKeydown"
                @paste="onPaste"
                @image-ready="imageReady"
                @image-added="imageAdded"
                @image-removed="imageRemoved"
                :disabled="disabled || waiting"
                :custom-translations="customTranslations"
                ref="input"
                :readonly="readonly"></froala>
    </m-input-style>

    <m-validation-message class="m-rich-text__validation-message"
                          :disabled="isDisabled"
                          :waiting="isWaiting"
                          :error="hasError"
                          :error-message="errorMessage"
                          :valid="isValid"
                          :valid-message="validMessage"
                          :helper-message="helperMessage"></m-validation-message>
</div>
