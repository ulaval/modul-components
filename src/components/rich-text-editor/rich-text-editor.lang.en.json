{
    "m-rich-text-editor": {
        "hide-submenu": "hide submenu",
        "insert": "insertions",
        "styles": "styles",
        "lists": "listes"
    }
}
