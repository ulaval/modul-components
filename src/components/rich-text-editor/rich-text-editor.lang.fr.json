{
    "m-rich-text-editor": {
        "hide-submenu": "Cacher le sous menu",
        "insert": "insertions",
        "styles": "styles",
        "lists": "listes"
    }
}
