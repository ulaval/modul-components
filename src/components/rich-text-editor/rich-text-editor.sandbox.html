<div>
    <!-- Sandbox template -->
    <h3>Froala Text Editor</h3>
    <h5 class="m-u--margin-bottom">fonctionne sans option spécifiée</h5>
    <div class="m-u--margin-bottom">
        <div><button @click="focus = !focus">Change focus (now: {{ focus }})</button></div>
        <div>
            Error message: <input type="text" v-model="errorMessage" />
        </div>
        <div>
            Valid message: <input type="text" v-model="validMessage" />
        </div>
        <div>
            Helper message: <input type="text" v-model="helperMessage" />
        </div>
        <div><button @click="waiting = !waiting">Waiting (now: {{ waiting }})</button></div>
        <div><button @click="disabled = !disabled">Disabled (now: {{ disabled }})</button></div>
    </div>
    <div class="m-u--margin-bottom">
        <m-rich-text-editor v-model="model"
        toolbar-sticky-offset=".mtest-header"
        :label="'texte riche'"
        :max-width="'large'"
        :placeholder="'some placeholder'"
        :focus="focus"
        :error="error"
        :error-message="errorMessage"
        :valid-message="validMessage"
        :helper-message="helperMessage"
        :waiting="waiting"
        :disabled="disabled"></m-rich-text-editor>
    </div>
    <div class="m-u--margin-bottom">
        <h5>Résultat dans le lecteur</h5>
        <div><m-rich-text :value="model"></m-rich-text></div>
    </div>
    <div>
        <h5 class="m-u--margin-bottom">fonctionne dans un formulaire pleine page</h5>
        <button @click="afficherFormulairePleinePage = !afficherFormulairePleinePage">Afficher formulaire pleine page</button>
        <m-edit-window :open.sync="afficherFormulairePleinePage">
            <template slot="header">
                <span class="m-u--h2 m-u--font-weight--bold">
                    <m-i18n k="texte:texte"></m-i18n>
                </span>
            </template>
            <div>
                <m-rich-text-editor scrollable-container=".m-edit-window__body" class="p-formulaire-texte__textearea" :label="'texte riche'"
                    tag-style="p" max-width="720px" v-model="fullScreenFormModel"></m-rich-text-editor>
            </div>
        </m-edit-window>
    </div>
    <h5 class="m-u--margin-bottom">placeholder vide n'affiche rien</h5>
    <div class="m-u--margin-bottom">
        <m-rich-text-editor v-model="model"
        :toolbar-sticky-offset="58"
        :label="'texte riche'"
        :max-width="'large'"
        :focus="focus"
        :error="error"
        :error-message="errorMessage"
        :valid-message="validMessage"
        :helper-message="helperMessage"
        :waiting="waiting"
        :disabled="disabled"></m-rich-text-editor>
    </div>
</div>
