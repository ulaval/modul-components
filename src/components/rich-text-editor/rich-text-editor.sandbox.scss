@import 'commons';

.m-rich-text-editor-sandbox {
    /deep/ .rte-h1 {
        @include m-h1-font-style;
        @include m-commons-title-styles;
    }

    /deep/ .rte-h2 {
        @include m-h2-font-style;
        @include m-commons-title-styles;
    }

    /deep/ .rte-h3 {
        @include m-h3-font-style;
        @include m-commons-title-styles;
    }

    /deep/ .rte-h4 {
        @include m-h4-font-style;
        @include m-commons-title-styles;
    }

    /deep/ .rte-h5 {
        @include m-h5-font-style;
        @include m-commons-title-styles;
    }

    /deep/ .rte-h6 {
        @include m-h6-font-style;
        @include m-commons-title-styles;
    }
}
