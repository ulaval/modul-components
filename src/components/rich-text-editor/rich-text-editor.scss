@import 'abstracts/variables';

.m-rich-text {
    position: relative;
    display: inline-flex;
    flex-direction: column;

    &.m--has-error {
        .m-rich-text__validation {
            color: $m-color--error;
        }
    }

    &.m--is-valid {
        .m-rich-text__validation {
            color: $m-color--success;
        }
    }

    &__validation {
        display: flex;
        justify-content: flex-end;

        &__message {
            flex: 1 1 auto;
        }
    }

    &__label {
        font-size: $m-font-size;
        font-weight: $m-font-weight--bold;
    }

    &__input {
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1;
        width: 100%;
        outline: none;
        resize: none;
        overflow: hidden;

        &-hidden {
            opacity: 0;
            word-wrap: break-word;
        }
    }
}
