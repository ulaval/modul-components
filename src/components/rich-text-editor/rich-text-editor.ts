import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';

import { ElementQueries } from '../../mixins/element-queries/element-queries';
import { InputLabel } from '../../mixins/input-label/input-label';
import { InputManagement, InputManagementData } from '../../mixins/input-management/input-management';
import { InputState, InputStateInputSelector } from '../../mixins/input-state/input-state';
import { InputWidth } from '../../mixins/input-width/input-width';
import uuid from '../../utils/uuid/uuid';
import { ModulVue } from '../../utils/vue/vue';
import { RICH_TEXT_EDITOR_NAME } from '../component-names';
import VueFroala from './adapter/vue-froala';
import { MRichTextEditorDefaultOptions, MRichTextEditorStandardOptions } from './rich-text-editor-options';
import WithRender from './rich-text-editor.html?style=./rich-text-editor.scss';

const RICH_TEXT_LICENSE_KEY: string = 'm-rich-text-license-key';

export enum MRichTextEditorMode {
    STANDARD
}

@WithRender
@Component({
    components: { Froala: VueFroala },
    mixins: [
        InputState,
        InputManagement,
        InputWidth,
        InputLabel,
        ElementQueries
    ]
})
export class MRichTextEditor extends ModulVue implements InputManagementData, InputStateInputSelector {
    selector: string = '.fr-element';
    internalValue: string;

    @Prop({ default: '' })
    public value: string;

    @Prop({
        default: MRichTextEditorMode.STANDARD,
        validator: value => value === MRichTextEditorMode.STANDARD
    })
    public mode: MRichTextEditorMode;

    @Prop({ default: '0' })
    public toolbarStickyOffset: string;

    @Prop()
    public scrollableContainer: string | undefined;

    protected id: string = `mrich-text-${uuid.generate()}`;

    mounted(): void {
        this.testSelectorProps();
    }

    public get internalOptions(): any {
        const propOptions: any = {
            // Hack to "hide" the default froala placeholder text
            placeholderText: this.as<InputManagement>()!.placeholder || ' ',
            toolbarStickyOffset: this.calculateToolbarStickyOffset(),
            scrollableContainer: this.getScrollableContainer()
        };

        return Object.assign(this.getDefaultOptions(), propOptions);
    }

    public get froalaLicenseKey(): string {
        return this.$license.getLicense<string>(RICH_TEXT_LICENSE_KEY) || '';
    }

    public getDefaultOptions(): MRichTextEditorDefaultOptions {
        if (this.mode === MRichTextEditorMode.STANDARD) {
            return new MRichTextEditorStandardOptions(this.froalaLicenseKey, this.$i18n.currentLang());
        }

        throw new Error(`rich-text-edit: mode ${this.mode} is not a valid mode.  See MRichTextEditMode Enum for a list of compatible modes.`);
    }

    public getSelectorErrorMsg(prop: string): string {
        return `${RICH_TEXT_EDITOR_NAME}: No element has been found with the selector given in the ${prop} prop.`;
    }

    protected refreshModel(newValue: string): void {
        this.$emit('input', newValue);
    }

    protected calculateToolbarStickyOffset(): number | undefined {
        if (this.toolbarStickyOffset) {
            if (/^\d+$/.test(this.toolbarStickyOffset)) {
                return Number(this.toolbarStickyOffset);
            } else {
                const element: HTMLElement | null = document.querySelector(this.toolbarStickyOffset);
                return element!.offsetHeight;
            }
        }
    }

    protected getScrollableContainer(): string | undefined {
        if (this.scrollableContainer && document.querySelector(this.scrollableContainer)) {
            return this.scrollableContainer;
        }
    }

    protected testSelectorProps(): void {
        let propInError: string | undefined;
        if (this.scrollableContainer) {
            if (!document.querySelector(this.scrollableContainer)) {
                propInError = 'scrollable-container';
            }
        }

        if (this.toolbarStickyOffset && !/^\d+$/.test(this.toolbarStickyOffset)) {
            if (!document.querySelector(this.toolbarStickyOffset)) {
                propInError = 'toolbar-sticky-offset';
            }
        }

        if (propInError) {
            console.error(this.getSelectorErrorMsg(propInError));
            throw new Error(this.getSelectorErrorMsg(propInError));
        }
    }
}
