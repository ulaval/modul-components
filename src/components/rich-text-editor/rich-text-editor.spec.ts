import { shallow, Wrapper } from '../../../node_modules/@vue/test-utils';
import Vue from '../../../node_modules/vue';
import { renderComponent } from '../../../tests/helpers/render';
import uuid from '../../utils/uuid/uuid';
import { MRichTextEditor, MRichTextEditorMode } from './rich-text-editor';
import { MRichTextEditorStandardOptions } from './rich-text-editor-options';
import RichTextLicensePlugin from './rich-text-license-plugin';

jest.mock('../../utils/uuid/uuid');
(uuid.generate as jest.Mock).mockReturnValue('uuid');

const froalaLicenseKey: string = 'testKey';
let wrapper: Wrapper<MRichTextEditor>;
let richTextEditor: MRichTextEditor;
let defaultOptions: MRichTextEditorStandardOptions;

describe('MRichTextEditor', () => {
    beforeEach(() => {
        Vue.use(RichTextLicensePlugin, { key: froalaLicenseKey });
        wrapper = shallow(MRichTextEditor);
        richTextEditor = wrapper.vm;
        defaultOptions = new MRichTextEditorStandardOptions(froalaLicenseKey, richTextEditor.$i18n.currentLang());
    });

    it('should render correctly', () => {
        expect(renderComponent(richTextEditor)).resolves.toMatchSnapshot();
    });

    it('the key is stored', () => {
        expect(richTextEditor.froalaLicenseKey).toEqual(froalaLicenseKey);
    });

    describe('In standard Mode', () => {
        beforeEach(() => {
            wrapper.setProps({
                mode: MRichTextEditorMode.STANDARD
            });
        });
        it('default options are standard default options', () => {
            expect(richTextEditor.getDefaultOptions()).toEqual(defaultOptions);
        });

    });

    describe('without props set', () => {
        it('internal options are defaultOptions', () => {
            const customOptions: any = { placeholderText: undefined, scrollableContainer: undefined };
            expect(richTextEditor.internalOptions).toEqual({ ...defaultOptions, ...customOptions });
        });
    });

    describe('with props set', () => {
        describe(`with props in error`, () => {
            it('the component should throw an error', () => {
                expect(() => {
                    wrapper = shallow(MRichTextEditor, {
                        propsData: { scrollableContainer: '#container' }
                    });
                }).toThrow(richTextEditor.getSelectorErrorMsg('scrollable-container'));

                expect(() => {
                    wrapper = shallow(MRichTextEditor, {
                        propsData: { toolbarStickyOffset: '#header' }
                    });
                }).toThrow(richTextEditor.getSelectorErrorMsg('toolbar-sticky-offset'));
            });
        });

        it('internal options are customed defaultOptions', () => {
            const customOptions: any = { toolbarStickyOffset: 1 };

            wrapper.setProps({ toolbarStickyOffset: 1 });
            expect(richTextEditor.internalOptions).toEqual({ ...defaultOptions, ...customOptions });
        });
    });
});
