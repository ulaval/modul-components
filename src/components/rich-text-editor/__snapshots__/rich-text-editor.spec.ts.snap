// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MRichTextEditor should render correctly 1`] = `
<div class="m-rich-text m--is-mobile" style="width:100%;max-width:288px;">
    <m-input-style label-for="mrich-text-uuid" empty="true" border-top="true" tag-style="default"></m-input-style>
    <div class="m-textfield__validation">
        <m-validation-message class="m-textfield__validation__message"></m-validation-message>
    </div>
</div>
`;
